File: 189. Rotate Array.py
from typing import List

class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        ##TLE
        # for i in range(k):
        #     for j in range(len(nums)):
        #         nums[j],nums[-1] = nums[-1],nums[j]
        def reverse(nums, l, r):
            while l < r:
                nums[l], nums[r] = nums[r], nums[l]
                l += 1
                r -= 1

        k = k % len(nums)  ##tricky
        reverse(nums, 0, len(nums) - 1)
        reverse(nums, 0, k - 1)
        reverse(nums, k, len(nums) - 1)

File: 189. Rotate Array.c
void reverse(int* nums, int begin, int end){
    if (begin >= end){
        return;
    }
    int tmp = *(nums+begin);
    *(nums+begin) = *(nums+end);
    *(nums+end) = tmp;
    
    reverse(nums, ++begin, --end);
}

void rotate(int* nums, int numsSize, int k){
    if (numsSize < 2) return;
    k %= numsSize;
    if (k==0) return;
    
    reverse(nums, 0, numsSize-k-1);
    // for (int i=0;i<numsSize;i++){
    //     printf("%d ", nums[i]);
    // }
    // printf("\n");
    reverse(nums, numsSize-k, numsSize-1);
    // for (int i=0;i<numsSize;i++){
    //     printf("%d ", nums[i]);
    // }
    // printf("\n");
    reverse(nums, 0, numsSize-1);
    // for (int i=0;i<numsSize;i++){
    //     printf("%d ", nums[i]);
    // }
    // printf("\n");
}
